packages/core/src/com/communication.ts

```diff
@@ -472,123 +479,41 @@ export class Communication {
         this.pendingEnvs.deleteKey(instanceId);
         delete this.environments[instanceId];
         for (const [callbackId, env] of this.callbackToEnvMapping.entries() ?? []) {
-            if (env === instanceId && this.callbacks[callbackId]) {
-                const { timerId, reject } = this.callbacks[callbackId]!;
-                clearTimeout(timerId);
-                reject(new Error(ENV_DISCONNECTED(instanceId)));
-                delete this.callbacks[callbackId];
-                this.callbackToEnvMapping.delete(callbackId);
+            const callbackRecord = this.callbacks[callbackId];
+            if (env === instanceId && callbackRecord) {
+                callbackRecord.reject(new Error(ENV_DISCONNECTED(instanceId, this.rootEnvId)));
             }
         }
         for (const dispose of this.disposeListeners) {
             dispose(instanceId);
         }
     }
 
-    private async forwardMessage(message: Message, env: EnvironmentRecord): Promise<void> {
-        const responseMessage: Message = {
-            from: message.to,
-            to: message.from,
-            callbackId: message.callbackId,
-            origin: message.to,
-            type: 'callback',
-            forwardingChain: [this.rootEnvId],
-        };
-
-        if (message.forwardingChain.indexOf(this.rootEnvId) > -1) {
+    private forwardMessage(message: Message, env: EnvironmentRecord) {
+        message.forwardingChain ??= [];
+        // check before push
+        const cycle = message.forwardingChain.includes(this.rootEnvId);
+        // push after check so we see the cycle in the forwardingChain
+        message.forwardingChain.push(this.rootEnvId);
+        if (cycle) {
+            console.error(FORWARDING_MESSAGE_STUCK_IN_CIRCULAR(cleanMessage(message), this.rootEnvId, env.id));
             this.sendTo(message.from, {
-                ...responseMessage,
+                from: this.rootEnvId,
+                origin: this.rootEnvId,
+                to: message.origin,
+                callbackId: message.callbackId,
+                type: 'callback',
+                forwardingChain: message.forwardingChain,
                 error: new Error(
                     `cannot reach environment '${message.to}' from '${message.from}' since it's stuck in circular messaging loop`,
```

Comment:
Move to logs

Reply:
removed the log from above and used the proper log message function